Refactor gradcheck

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54480 Implement faster gradcheck and enable for all ops except some
* #54479 Factor out numerical logic
* #54378 Refactor numerical to calculate wrt all outputs at once
* #54092 Refactor get numerical jacobian
* #54049 Refactor get analytical jacobian
* #53916 Gradcheck small fixes
* **#53857 Refactor gradcheck**

This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - ~when inputs are not dense, but check_sparse_nnz is false~
  - ~when none of the inputs require grad~
  - ~(warning) when inputs are not double precision~
  - ~when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.~
- check_no_differentiable_outputs:
  - ~when none of the outputs are differentiable, but numerical gradient is not zero~
- check_outputs:
  - ~when sparse outputs (always raise)~
  - ~when mkldnn outputs (always raise)~
- test_batched_grad
  - ~when encounter runtime error while computing batched grad (print big message)~
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - ~when layout of grad_input is not the same as input~
  - ~when grad_input is sparse and has incorrect sparse_dim/dense_dim~
  - ~when backward not multiplied by grad_output (sparse/non-sparse case)~
  - when grad is incorrect type/size
- test_undefined_grad
  - ~when encounter runtime error while running backward~
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size

Differential Revision: [D27201571](https://our.internmc.facebook.com/intern/diff/D27201571)

soulitzer committed:
title

[ghstack-poisoned]

soulitzer committed:
Update on "[Do not review] testing"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor gradcheck"




[ghstack-poisoned]

soulitzer commented:
This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions

soulitzer committed:
Update on "(WIP) Refactor gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "(WIP) Refactor gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"




[ghstack-poisoned]

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -160,18 +161,41 @@ def get_stride(size):
+            raise RuntimeError(msg)
+
+    if not correct_grad_types and check_grad_dtypes:
+        fail_test('Gradient{0} has dtype mismatch'.format(complex_str))
```
nit: We can use f-strings to make this more concise. e.g. `fail_test(f'Gradient{complex_str} has dtype mismatch')`. Ditto for the other strings in this file.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -160,18 +161,41 @@ def get_stride(size):
-    diff_input_list = list(iter_tensors(input, True))
-    jacobian = make_jacobian(input, output.numel())
-    jacobian_reentrant = make_jacobian(input, output.numel())
+def check_analytical_jacobian(inputs, output, nondet_tol, grad_out, check_batched_grad,
+                              check_grad_dtypes, raise_exception):
```
nit: If we're refactoring this anyways, it would be nice to add type hints (either now or in the future). For example, it's important to note here that inputs is a `Tuple` (I think? At least from reading the original code this seemed to be the case)

soulitzer replied:
mypy has been a little tricky, lets save this for the future

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
-    for i, o in enumerate(output):
-        if not o.requires_grad:
-            continue
+    check_outputs(outputs)
```
Do we have tests for gradcheck? I noticed some logic being moved around (like `check_outputs`).

soulitzer replied:
Maybe there are some scattered around - definitely not enough though. For example `raise_exception=False` isn't tested at all.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
-        analytical, reentrant, correct_grad_sizes, correct_grad_types = get_analytical_jacobian(tupled_inputs,
-                                                                                                o,
-                                                                                                nondet_tol=nondet_tol)
+        analytical, failed = check_analytical_jacobian(tupled_inputs, o, nondet_tol, 1.0, check_batched_grad,
```
The name here is a bit ambiguous. What is this actually checking? Maybe "check_analytical_jacobian_attributes" makes more sense (because the check runs tests for dtypes and sizes among other things)

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
-            return fail_test('Analytical gradient (calculated using complex valued grad output) has incorrect size')
+        if o.is_complex():
+            analytical_imag_grad_out, failed = check_analytical_jacobian(tupled_inputs, o, nondet_tol, 1j,
+                                                                         check_batched_grad, check_grad_dtypes,
```
It feels a little weird to me that we have to pass in `check_batched_grad` to `check_analytical_jacobian`. Ideally all the batched grad logic would be together. Could we have `check_analytical_jacobian` return `reentrant`?

soulitzer replied:
Hmm, we could do that. But now I'm thinking that we could remove the second batched-grad specific `reentrant` check entirely since we've simplified the main function so much, it seems clear that `failed` would be False when `reentrant` is false, and we'd always early exit in that case.

zou3519 replied:
Yeah, I'm fine with removing the batched-grad `reentrant` assert.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
+                                                                         raise_exception)
+            if failed:
+                return False
+            numerical_imag_grad_out = get_numerical_jacobian(fn, tupled_inputs, eps=eps, grad_out=1j)
```
Does `checkIfNumericalAnalyticAreClose` get used at all in the new code? If not, this probably means we don't have any gradcheck tests lol.

soulitzer replied:
Its no longer used in the next PR in the stack, but it is still used in this one :468 for example. Indeed there is no gradcheck test catching the fact that we don't actually return the result of `checkIfNumericalAnalyticAreClose` (similarly for `test_batched_grad`), but I choose to add the return and changes nearby to that PR so this one could ideally have no behavior change.

zou3519 replied:
Oh, *facepalm, the way the blame shows up I read it as like 468 gets called from inside `checkIfNumericAnalyticAreClose`. Thanks for clarifying this, that makes sense

zou3519 reviewed (commented):
Some questions. The general flow of this looks good. I had some high-level comments:
- do we have any tests for gradcheck?
- In the future it would be good for the PR body to include more rationale. It looks like the rationale for this refactor is to make the logic directly in the `gradcheck` function easier to read, is that all?
- I'm not sure that after the refactor, we're calling `checkIfNumericalAnalyticAreClose`, but maybe I am not looking in the right place

soulitzer committed:
Update on "Refactor gradcheck"




[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer commented:
@zou3519
> do we have any tests for gradcheck?

I've updated the PR body to include some discussion of tests (some have also been added in the last update). 
> In the future it would be good for the PR body to include more rationale. It looks like the rationale for this refactor is to make the logic directly in the gradcheck function easier to read, is that all?

Yep, the rationale is like you said, to make it easier to read, but also to avoid code duplication when we introduce fast gradcheck.

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -455,9 +575,9 @@ def checkIfNumericalAnalyticAreClose(a, n, j, error_str=''):
-                    a_with_imag_grad_out = analytical_with_imag_grad_out[j]
-                    n_with_imag_grad_out = numerical_with_imag_grad_out[j]
-                    checkIfNumericalAnalyticAreClose(a_with_imag_grad_out, n_with_imag_grad_out, j,
+                    a_imag_grad_out = analytical_imag_grad_out[j]
+                    n_imag_grad_out = numerical_imag_grad_out[j]
```
I don't understand these variable names, is the implication that the jacobian is full of imaginary values?

soulitzer replied:
Yeah its a bit confusing. What its supposed to mean is "analytical [jacobian calculated using] imag. grad out" lol

zou3519 replied:
analytical_calculated_using_imag_grad_out is actually not that long/bad of a name lol

soulitzer replied:
I'm just going to say "analytical_from_imag_grad_out"

zou3519 commented on torch/autograd/gradcheck.py:
```diff
@@ -279,6 +360,93 @@ def vjp(v):
+
+
+def test_undefined_grad(fail_test, func, inputs) -> bool:
+    output = _differentiable_outputs(func(*inputs))
```
Previously, we were computing `func(*input)` once (see previous like 498). Now, we're doing it twice, once in test_undefined_grad and once in test_backward_mul_by_grad_output.

I assume this is fine w.r.t performance? The bulk of the work in gradcheck is probably the numeric jacobian computation.

soulitzer replied:
I think it might be because can't backward through the same graph multiple times (when there are saved variables that are needed for backward).

soulitzer replied:
This is a good catch, and it should be fixed now. Numeric jacobian is indeed the bulk of the computation right now, but only because its computing func so many times. Currently, we call func at least 2 * num_outputs * num_inputs * dim_inputs  times to compute numerical jacobian (there might be a way to get it to 2 * num_outputs * dim_inputs, but requires more extensive refactoring). So, though it might seem like a relatively small improvement right now, I think this could be significant especially once we use it in the fast-gradcheck version, which only needs to call func 2 * num_inputs times to compute numerical jacobian.

zou3519 reviewed (commented):
The refactor LGTM, I'm happy to stamp to unblock if we want to push writing tests to a future change. On the other hand, writing the tests now will give us more confidence that we aren't breaking anything. Thanks for the descriptive PR body, it helps a lot for reviewing!

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad (print big message)
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad (print big message)
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad (print big message)
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - when inputs are not dense, but check_sparse_nnz is false
  - when none of the inputs require grad
  - (warning) when inputs are not double precision
  - when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.
- check_no_differentiable_outputs:
  - when none of the outputs are differentiable, but numerical gradient is not zero
- check_outputs:
  - when sparse outputs (always raise)
  - when mkldnn outputs (always raise)
- test_batched_grad
  - when encounter runtime error while computing batched grad (print big message)
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - when no Tensors requiring grad found in input
  - when layout of grad_input is not the same as input
  - when grad_input is sparse and has incorrect sparse_dim/dense_dim
  - when backward not multiplied by grad_output (sparse/non-sparse case)
  - when grad is incorrect type/size
- test_undefined_grad
  - when no Tensors requiring grad found in input
  - when encounter runtime error while running backward
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size
  - when not reentrant (DONE)

[ghstack-poisoned]

soulitzer commented:
@zou3519 
I think I feel safe enough with the number of tests we have introduced in this PR, since its mainly moving big pieces around, and the tests we introduced does cover the major failure cases in each function.

So if you think it looks, good we can go ahead and ship!

zou3519 commented on test/test_autograd.py:
```diff
@@ -3965,6 +3955,131 @@ def backward(ctx, grad_out):
+        # TODO: when we complete backward but grad inputs (the output of .grad()) is not none
+
+    def test_gradcheck_check_analytical_jacobian_attributes(self):
+        # TODO: when grad_input is incorrect dtype/size
```
Let's delete this test and add an issue for this? Or add an issue and link it from the TODO. Ditto for the other TODOs that this PR adds.

soulitzer replied:
Tracking issue for missing tests created: https://github.com/pytorch/pytorch/issues/54346

zou3519 commented on test/test_autograd.py:
```diff
@@ -3965,6 +3955,131 @@ def backward(ctx, grad_out):
         gradcheck(lambda x: NonDetFunc.apply(x, 1e-6), inp, nondet_tol=1e-5, check_batched_grad=False)
         gradgradcheck(lambda x: NonDetFunc.apply(x, 1e-12), inp, nondet_tol=1e-5, check_batched_grad=False)
 
+    def test_gradcheck_test_inputs(self):
```
nit: "validates inputs" or "validates test inputs"  is more descriptive than "test inputs"

zou3519 commented on test/test_autograd.py:
```diff
@@ -3965,6 +3955,131 @@ def backward(ctx, grad_out):
+        # when inputs are not dense, but check_sparse_nnz is false
+        x = torch.rand(10, requires_grad=True).to_sparse()
+        with self.assertRaisesRegex(RuntimeError, 'dense when check_sparse_nnz is set to False.'):
+            gradcheck(lambda x: x.to_dense(), (x,), check_sparse_nnz=False, check_batched_grad=False)
```
Should we be careful that we're testing the public gradcheck API here (torch.autograd.gradcheck)? I think right now we're using the private gradcheck API.

For the purposes of this test it looks like it doesn't actually matter, but if we diverge the behavior of the public and testing-only gradcheck more then it may matter.

soulitzer replied:
Good point. we can revisit this when we start to think about how we test that slow/fast have the same behavior.

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - ~when inputs are not dense, but check_sparse_nnz is false~
  - ~when none of the inputs require grad~
  - ~(warning) when inputs are not double precision~
  - ~when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.~
- check_no_differentiable_outputs:
  - ~when none of the outputs are differentiable, but numerical gradient is not zero~
- check_outputs:
  - ~when sparse outputs (always raise)~
  - ~when mkldnn outputs (always raise)~
- test_batched_grad
  - ~when encounter runtime error while computing batched grad (print big message)~
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - ~when layout of grad_input is not the same as input~
  - ~when grad_input is sparse and has incorrect sparse_dim/dense_dim~
  - ~when backward not multiplied by grad_output (sparse/non-sparse case)~
  - when grad is incorrect type/size
- test_undefined_grad
  - ~when encounter runtime error while running backward~
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size

[ghstack-poisoned]

soulitzer committed:
Update on "Refactor gradcheck"


This PR basically just factors a lot of the logic out from the main gradcheck function into their own individual functions. It aims to avoid any behavior change (but we may not have enough tests to actually verify this). Refactorings that lead to any behavior chang are done in the next PR in this stack. 

The rationale for this change is 1) to make the main gradcheck function cleaner to read, and 2) also allow us to reuse the same pieces when we add the fast gradcheck.

Maybe this PR is also a good place to add some tests for gradcheck, i.e., make sure gradcheck fails when it should fail, as to make sure that we are indeed not changing any logic. This will also help us make sure our fast_gradcheck does all the necessary checks:
So far existing tests are: 
- test_gradcheck_fail_when_no_differentiable_outputs_and_num_grad_not_zero` (test_autograd)
- test_gradcheck_single_input (test_autograd)
- test_gradcheck_sparse_input (test_autograd)
- test_gradcheck_nondeterministic (test_autograd)
- test_gradcheck (test_overrides)

Full coverage would potentially require adding the following missing tests (for each test for both raise_exception=True/False) - Methodology for getting the list below is that for every type of error message we spit out, we make sure we can hit it:
- complex:
  - when numerical != analytical when tested with imag grad_out
- check_inputs
  - ~when inputs are not dense, but check_sparse_nnz is false~
  - ~when none of the inputs require grad~
  - ~(warning) when inputs are not double precision~
  - ~when layout is not mkldnn(aka has strides) and input has a dimension with stride 0.~
- check_no_differentiable_outputs:
  - ~when none of the outputs are differentiable, but numerical gradient is not zero~
- check_outputs:
  - ~when sparse outputs (always raise)~
  - ~when mkldnn outputs (always raise)~
- test_batched_grad
  - ~when encounter runtime error while computing batched grad (print big message)~
  - when not allclose (print out big message)
- test_backward_mul_by_grad_output
  - ~when layout of grad_input is not the same as input~
  - ~when grad_input is sparse and has incorrect sparse_dim/dense_dim~
  - ~when backward not multiplied by grad_output (sparse/non-sparse case)~
  - when grad is incorrect type/size
- test_undefined_grad
  - ~when encounter runtime error while running backward~
  - when we complete backward but grad inputs (the output of .grad()) is not none 
- check_analytical_jacobian_attributes (for both complex/non complex)
  - when grad input is incorrect dtype/size

Differential Revision: [D27201571](https://our.internmc.facebook.com/intern/diff/D27201571)

[ghstack-poisoned]
